pkg/thanos/rules.go

```diff
@@ -0,0 +1,334 @@
+// Copyright 2016 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+
+	monitoringv1 "github.com/coreos/prometheus-operator/pkg/apis/monitoring/v1"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/ghodss/yaml"
+	"github.com/go-kit/kit/log/level"
+	"github.com/openshift/prom-label-proxy/injectproxy"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/promql"
+)
+
+const labelPrometheusName = "prometheus-name"
+
+// The maximum `Data` size of a ConfigMap seems to differ between
+// environments. This is probably due to different meta data sizes which count
+// into the overall maximum size of a ConfigMap. Thereby lets leave a
+// large buffer.
+var maxConfigMapDataSize = int(float64(v1.MaxSecretSize) * 0.5)
+
+func (o *Operator) createOrUpdateRuleConfigMaps(t *monitoringv1.ThanosRuler) ([]string, error) {
```

Comment:
How come we redefine this here? Is this not reusable from the Prometheus flow that does the same thing? I'd expect the Thanos ruler and Prometheus to use shared code for this (maybe this is simply a todo, in which case ignore this comment).

Reply:
I didn't reuse it directly at first because it uses several fields from the Prometheus spec, but it seems the logic is the same other than the names of things.  So I think you are right that it should be reusable after refactoring to make the original functions a bit more generic.

Reply:
Yes I assume something to that regard, makes perfect sense!

Reply:
I would still like to see this before we merge.